src/Processors/Formats/Impl/AvroRowInputFormat.cpp

```diff
@@ -438,6 +434,80 @@ AvroDeserializer::DeserializeFn AvroDeserializer::createDeserializeFn(const avro
                         target_type->getName());
                 }
             }
+
+            if (target.isVariant())
+            {
+                const DataTypeVariant & variant_type = assert_cast<const DataTypeVariant &>(*target_type);
+                const auto & nested_types = variant_type.getVariants();
+
+                std::vector<DeserializeFn> nested_deserializers;
+                nested_deserializers.reserve(root_node->leaves());
+
+                size_t avro_union_null_index = -1;
+                for (size_t i = 0; i != root_node->leaves(); ++i)
+                {
+                    const auto & avro_node = root_node->leafAt(static_cast<int>(i));
+                    if (avro_node->type() == avro::AVRO_NULL)
+                    {
+                        avro_union_null_index = i;
+                        continue;
+                    }
+                    nested_deserializers.emplace_back(createDeserializeFn(avro_node, nested_types.at(i)));
+                }
```

Comment:
Unfortunately it won't work like this. `Variant` type doesn't save the order of the variant types provided during creation. We consider `Variant(T1, T2) = Variant(T2, T1)`, so inside `DataTypeVariant` we store types in sorted by type name order.
So, here `nested_types[i]` won't correspond to `root_node->leafAt(i)`. We should somehow create a mapping from `nested_types` to types inside `union` node.

Reply:
The simpliest way is to use `avroNodeToDataType` for each type inside `union` and try to find the same type in variants.

Reply:
Thanks for pointing that out! I have just reflected your finding in the latest version of this PR. Can you please take another look?